File: SkiCameraAndroid/shared/src/commonMain/kotlin/com/vcnt/skicamera/shared/PIDController.kt
package com.vcnt.skicamera.shared

import kotlin.math.sqrt

/**
 * A standard PD Controller (Integral is omitted for auto-zoom stability).
 */
class PIDController(
    var kp: Double,
    var kd: Double
) {
    private var prevError: Double = 0.0

    /**
     * Calculates the control output (velocity).
     * @param error The difference between Target and Current.
     * @param dt Delta time in seconds.
     */
    fun update(error: Double, dt: Double): Double {
        if (dt <= 0) return 0.0

        // Derivative term: Rate of change of error
        val derivative = (error - prevError) / dt
        prevError = error

        return (kp * error) + (kd * derivative)
    }

    fun reset() {
        prevError = 0.0
    }
}

File: SkiCameraIOS/SkiCameraIOS/Camera/Logic/PIDController.swift
import Foundation

class PIDController {
    var kp: Double
    var kd: Double
    private var lastError: Double = 0.0
    
    init(kp: Double, kd: Double) {
        self.kp = kp
        self.kd = kd
    }
    
    func update(_ error: Double, _ dt: Double) -> Double {
        if dt <= 0 { return 0.0 }
        let derivative = (error - lastError) / dt
        lastError = error
        return (kp * error) + (kd * derivative)
    }
    
    func reset() {
        lastError = 0.0
    }
}
